src/gmcp.cpp

```diff
@@ -0,0 +1,189 @@
+            !IS_SET(EXIT(ch, door)->exit_info, EX_HIDDEN))
+          j["exits"].push_back({"direction", exitdirs[door], "state", "INACCESSIBLE"});
+        else if (!IS_SET(EXIT(ch, door)->exit_info, EX_CLOSED | EX_HIDDEN))
+          j["exits"].push_back({"direction", exitdirs[door], "state", "OPEN"});
+      } else {
+        if (!IS_SET(EXIT(ch, door)->exit_info, EX_HIDDEN) || GET_LEVEL(ch) > LVL_MORTAL) {
+          if (IS_SET(EXIT(ch, door)->exit_info, EX_LOCKED))
+            j["exits"].push_back({"direction", exitdirs[door], "state", "LOCKED"});
+          else if (IS_SET(EXIT(ch, door)->exit_info, EX_CLOSED))
+            j["exits"].push_back({"direction", exitdirs[door], "state", "CLOSED"});
+          else
+            j["exits"].push_back({"direction", exitdirs[door], "state", "OPEN"});
+        }
+      }
+    }
+
+  // Dump the json to a string and send it.
+  std::string payload = j.dump();
+  SendGMCP(ch->desc, "Room.Exits", payload.c_str());
+}
+
+void SendGMCPRoomInfo( struct char_data *ch, struct room_data *room ) 
+{
+  if (!ch || !ch->desc || !ch->desc->pProtocol->bGMCP) return;
+  json j;
+
+  j["room_vnum"] = GET_ROOM_VNUM(room);
+  j["room_name"] = GET_ROOM_NAME(room);
+  j["zone_number"] = room->zone;
+  // Only add coordinates if they are valid.
+  if (room->x && room->y && room->z) {
+    j["coords"] = { {"x", room->x}, {"y", room->y}, {"z", room->z} };
+  }
+
+  // Add room description if it exists; any quotes will be automatically escaped.
+  if (*room->description) {
+    j["room_description"] = room->description;
+  }
+
+  if (room->latitude) j["latitude"] = room->latitude;
+  if (room->longitude) j["longitude"] = room->longitude;
+  
+  // Dump the json to a string and send it.
+  std::string payload = j.dump();
+  SendGMCP(ch->desc, "Room.Info", payload.c_str());
+
+  SendGMCPExitsInfo(ch);
+}
+
+void SendGMCP( descriptor_t *apDescriptor, const char *module, const char *apData )
+{
+  if (!apDescriptor->pProtocol->bGMCP) return;
+  char buf[MAX_STRING_LENGTH];
+  // Build the GMCP message: IAC SB GMCP [module] [json] IAC SE
+  snprintf(buf, sizeof(buf), "%c%c%c%s %s%c%c", IAC, SB, TELOPT_GMCP, module, apData, IAC, SE);
+
+  Write(apDescriptor, buf);
+}
+
+void SendGMCPCharStatus( struct char_data * ch )
+{
+  if (!ch || !ch->desc || !ch->desc->pProtocol->bGMCP) return;
+  json j;
+
+  j["physical"] = GET_PHYSICAL(ch);
+  j["mental"] = GET_MENTAL(ch);
+  j["physical_max"] = GET_MAX_PHYSICAL(ch);
+  j["mental_max"] = GET_MAX_MENTAL(ch);
+  j["pools"] = {"body", GET_BODY_POOL(ch), "magic", GET_MAGIC_POOL(ch), "combat", GET_COMBAT_POOL(ch), "hacking", GET_HACKING(ch)};
+  j["task_pools"] = json::object();
+  for (int x = 0; x < 7; x++)
+    j["task_pools"][attributes[x]] = GET_TASK_POOL(ch, x);  
+
+  // Dump the json to a string and send it.
+  std::string payload = j.dump();
+  SendGMCP(ch->desc, "Char.Status", payload.c_str());
+}
+
+void SendGMCPCharPools( struct char_data * ch )
```

Comment:
This seems like duplication of the code from above. Should the pools JSON composition logic be moved to a function that's called from both of these?